File: column/c/column.c
#include <stdio.h>
#include <stdlib.h>

#define IS_EMPTY(X) ((X) == NULL)

typedef struct r
{
    int data;
    struct r *up;
    struct r *down;
} row;

typedef row * rowPtr;

typedef struct c
{   
    char label; 
    struct c *next;
    struct c *prev;
    rowPtr row;
} column;

typedef column * columnPtr;

void help()
{
    printf("Enter number 1 to add a new item\n");
    printf("Enter number 3 to print your Stack\n");
    printf("Enter number 4 to Exit\n");
}

int main()
{
    columnPtr my_c = malloc(sizeof(column));

    if (IS_EMPTY(my_c))
        return 1;

    my_c->label = 'A';

    add_column(&my_c, 'B');

    int choice;

    while (1)
    {
        help();
        scanf("%d", &choice);
        switch (choice)
        {
            case 1:
                int item;
                printf("Entrt Your Item to Add: ");
                scanf("%d", &item);
                
                if (item%2 == 0)
                {
                    if (my_c->label == 'B')
                        my_c = my_c->prev;
                }
                else
                {
                    if (my_c->label == 'A')
                        my_c = my_c->next;
                }

                add_data_to_column(item, &my_c);
                printf("\n\n\n\n\n");
                break;
            case 3:
                printf("colcumn : \n");
                print_all_column(&my_c);
                printf("\n\n\n\n\n");
                break;
            default:
                return 0;
                break;
        }
    }
    
}

void add_column(columnPtr *col, char label)
{
    columnPtr new_column = malloc(sizeof(column));


    if (!IS_EMPTY(new_column))
    {   
        new_column->label = label;
        
        if (IS_EMPTY((*col)->next))
        {
            (*col)->next = new_column;
            new_column->prev = (*col); 
        }
        else
        {   
            columnPtr courrent_column = (*col)->next;
            while (!IS_EMPTY(courrent_column->next))
            {
                courrent_column = courrent_column->next;
            }
            
            courrent_column->next = new_column;
            new_column->prev = courrent_column;
        }
    }
}

void add_data_to_column(int data, columnPtr *col)
{
    rowPtr my_r = malloc(sizeof(row));
    if (!IS_EMPTY(my_r))
    {   
        my_r->data = data;

        if (IS_EMPTY((*col)->row))
        {
            (*col)->row = my_r;
        }
        else 
        {   
            rowPtr prev_row = (*col)->row;
            while (prev_row->data < my_r->data && !IS_EMPTY(prev_row->down))
            {
                prev_row = prev_row->down;
            }
            prev_row->down = my_r;
            my_r->up = prev_row;
        }

    }
}

void print_all_column(columnPtr *col)
{
    
    columnPtr curent_c = *col;
    rowPtr curent_r;

    while (!IS_EMPTY(curent_c))
    {   
        printf("%c:\n", curent_c->label);
        curent_r = curent_c->row;
            while (!IS_EMPTY(curent_r))
            {
                printf("%d\n", curent_r->data);
                curent_r = curent_r->down;
            }
        printf("\n");
        curent_c = curent_c->next;
    }

    printf("\n");

}

File: column/Python/column.py
class Row:
    item = None
    down = None
    up = None

    def __init__(self, item = 0, down = None, up = None):
        self.item = item
        self.down = down
        self.up = up


class Column:
    label = ""
    next = None
    prev = None
    row: Row = None

    def __init__(self, label: str = "Column", next = None, prev = None, row: Row = None):
        self.label = label
        self.next = next
        self.prev = prev
        self.row = row

def add_column(column: Column, label: str):

    new_column = Column(label = label)

    if column.next is None:
        column.next = new_column
        new_column.prev = column

    elif curent_column.next is not None:
        
        while curent_column.next is not None:
            curent_column = curent_column.next
        
        curent_column.next = new_column
        new_column.prev = curent_column
    

def add_data_to_column(column: Column, item):
    
    new_row = Row(item=item)

    if column.row is None:
        column.row = new_row

    else:
        row = column.row
        while row.down is not None:
            row = column.row.down

        row.down = new_row
        new_row.up = row


def print_all_column(column: Column):

    c = column
    r :Row

    while c is not None:
        print(f"{c.label} :")
        r = column.row

        while r is not None:
            print(f"{r.item}")
            r = r.down

        c = c.next
            

def help():
    print("Enter number 1 to add a new item")
    print("Enter number 3 to print All Columns")
    print("Enter number 4 to Exit")

if __name__ == "__main__":

    c = Column("A")
    add_column(c, "B")

    is_run = True

    while is_run:
        help()

        select = int(input())

        if select == 1:
            item = int(input("Enter Your Number: "))
            add_data_to_column(c, item)

        if select == 3:
            print_all_column(c)
        
        if select == 4:
            is_run = False
        
        if select > 4 or select < 1:
            print(f"Number {select} is Not In list! please select correct number\n\n")
            continue

